app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java

```diff
@@ -2,27 +2,54 @@
 
 import android.os.Bundle;
 import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
 
+import com.vividsolutions.jts.geom.Envelope;
+import com.vividsolutions.jts.geom.Geometry;
+import com.vividsolutions.jts.geom.Point;
+
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 
 import javax.inject.Inject;
 
+import de.westnordost.osmapi.map.data.BoundingBox;
+import de.westnordost.osmapi.map.data.Element;
 import de.westnordost.streetcomplete.R;
-import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
+import de.westnordost.streetcomplete.data.osm.Countries;
+import de.westnordost.streetcomplete.data.osm.OsmElementQuestType;
 import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
+import de.westnordost.streetcomplete.data.osm.download.MapDataWithGeometryHandler;
 import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
+import de.westnordost.streetcomplete.data.osm.tql.FiltersParser;
 import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;
 import de.westnordost.streetcomplete.quests.YesNoQuestAnswerFragment;
+import de.westnordost.streetcomplete.quests.housenumber.AddHousenumber;
+import de.westnordost.streetcomplete.util.JTSConst;
 
-public class AddToiletAvailability extends SimpleOverpassQuestType
+public class AddToiletAvailability implements OsmElementQuestType
 {
-	@Inject public AddToiletAvailability(OverpassMapDataDao overpassServer) { super(overpassServer); }
 
-	@Override protected String getTagFilters()
+	private final OverpassMapDataDao overpassServer;
+
+	@Inject public AddToiletAvailability(OverpassMapDataDao overpassServer)
+	{
+		this.overpassServer = overpassServer;
+	}
+
+	/** Query that returns all elements that don't have a toilet tag set */
+	private static String getElementsWithoutToiletsOverpassQuery(BoundingBox bbox)
 	{
 		// only for malls and big stores because users should not need to go inside a non-public
 		// place to solve the quest. (Considering malls and department stores public enough)
-		return "nodes, ways with shop ~ mall|department_store and name and !toilets";
+		return new FiltersParser().parse("nodes, ways with (shop ~ mall|department_store and name and !toilets) or (highway=rest_area and !toilets)").toOverpassQLString(bbox);
+	}
+
+	/** Query that returns all toilets */
+	private static String getFreeFloatingToiletsOverpassQuery(BoundingBox bbox)
```

Comment:
is it possible to filter toilets inside in overpass api?

Reply:
Not that I know of... It is done the same with the housenumber quest like with this quest...